PublicGitArchive/pga/cmd/cache.go

```diff
@@ -17,32 +18,77 @@ const (
 
 // updateCache checks whether a new version of the file in url exists and downloads it
 // to dest. It returns an error when it was not possible to update it.
-func updateCache(dest, source FileSystem, name string) error {
+func updateCache(ctx context.Context, dest, source FileSystem, name string) error {
 	logrus.Debugf("syncing %s to %s", source.Abs(name), dest.Abs(name))
 	if upToDate(dest, source, name) {
 		logrus.Debugf("local copy is up to date")
 		return nil
 	}
 
 	logrus.Debugf("local copy is outdated or non existent")
-	wc, err := dest.Create(name)
+	tmpName := name + ".tmp"
+	if err := copy(ctx, source, dest, name, tmpName); err != nil {
+		if cerr := dest.Remove(tmpName); cerr != nil {
+			logrus.Warningf("error removing temporary file %s: %v",
+				dest.Abs(tmpName), cerr)
+		}
+
+		return fmt.Errorf("could not copy to temporary file %s: %v",
+			dest.Abs(tmpName), err)
+	}
+
+	if err := dest.Rename(tmpName, name); err != nil {
+		return fmt.Errorf("rename %s to %s failed: %v",
+			dest.Abs(tmpName), dest.Abs(name), err)
+	}
+
+	return nil
+}
+
+func copy(ctx context.Context, source, dest FileSystem,
+	sourceName, destName string) (err error) {
+
+	wc, err := dest.Create(destName)
 	if err != nil {
-		return fmt.Errorf("could not create %s: %v", dest.Abs(name), err)
+		return fmt.Errorf("could not create %s: %v", dest.Abs(destName), err)
 	}
+	defer checkClose(dest.Abs(destName), wc, &err)
 
-	rc, err := source.Open(name)
+	rc, err := source.Open(sourceName)
 	if err != nil {
 		return err
 	}
-	defer rc.Close()
+	defer checkClose(source.Abs(sourceName), rc, &err)
 
-	if _, err = io.Copy(wc, rc); err != nil {
-		return fmt.Errorf("could not copy %s to %s: %v", source.Abs(name), dest.Abs(name), err)
+	if _, err = cancelableCopy(ctx, wc, rc); err != nil {
+		return fmt.Errorf("could not copy %s to %s: %v",
+			source.Abs(sourceName), dest.Abs(destName), err)
 	}
-	if err := wc.Close(); err != nil {
-		return fmt.Errorf("could not close %s: %v", dest.Abs(name), err)
+
+	return err
+}
+
+func cancelableCopy(ctx context.Context, dst io.Writer, src io.Reader) (
+	int64, error) {
+
+	var written int64
+	for {
+		select {
+		case <-ctx.Done():
+			return written, fmt.Errorf("download interrupted")
+		default:
+		}
+
+		w, err := io.CopyN(dst, src, 512*1024)
```

Comment:
why 512*1024? should this be a named constant?

Reply:
Yes, that was an arbitrary number for testing. I'll change it.